### PR TITLE
Relation support in graphback

### LINKED ISSUE #130: Relationship support in schema and resolvers
## Background
`graphback` version `0.3.1` creates database structure for relationships.

## TODO
Add relationships in `schema` and `resolvers`
For a model
```
type Note {
    id: ID!
    title: String!
    comments: [Comment!]!
}

type Comment {
    id: ID!
    content: String!
}
```
we can create the schema with types like
```
type Note {
    id: ID!
    title: String!
    comments: [Comment]!
}

type Comment {
    id: ID!
    content: String!
    noteId: ID!
}
```
Not sure how to work with n:m relations for now
Resolvers can be implemented using either
- Dataloader
- Graphql-query-mapper

@wtrocki What should be the methods we should provide for relationships? like `findXsOfY` for example.

### PR DESCRIPTION
**Pull request by ankitjena**
## Work done
Added relationships info to schema generation

### Example
```
type Note {
  id: ID!
  title: String!
  description: String!
  ## Relationship
  comment: [Comment]!
}

type Comment {
  id: ID!
  title: String!
  description: String!
}
```
output schema looks like
```
type Note {
  id: ID!
  title: String!
  description: String!
  comment: [Comment]!
}

type Comment {
  id: ID!
  title: String!
  description: String!
  noteId: ID!
}

input NoteInput {
  title: String!
  description: String!
}

input CommentInput {
  title: String!
  description: String!
  noteId: ID!
}

input NoteFilter {
  id: ID
  title: String
  description: String
}

input CommentFilter {
  id: ID
  title: String
  description: String
  noteId: ID
}
......
```
Changes in output resolvers
```
Note: {
  comments: (parent: any, _: any, context: GraphQLContext) => {
    return context.db.select().from('comment').where('noteId', '=', parent.id)
  }
},
......
```
Resolves #130

### CODE REVIEW

**Comment by wtrocki** [comment]
Is this PR going to have implementation for resolvers

**Comment by ankitjena** [comment]
> Is this PR going to have implementation for resolvers

Yes

File: packages/graphback/src/resolvers/resolverTemplate.ts [hunk 1]
```diff
@@ -3,8 +3,32 @@ import { TargetResolverContext } from './knex/targetResolverContext';
...
-  return `${imports}
+  if(context.relations.length) {
+    return `${imports}
+  
+enum Subscriptions {
+  ${context.subscriptionTypes.join(',\n  ')}
+}
+
+export const resolvers = {
+  ${context.relations.join(',\n\n  ')},
```
Comment (wtrocki) [inline packages/graphback/src/resolvers/resolverTemplate.ts]: I would see them on bottom actually

**Comment by wtrocki** [comment]
>    return context.db.select().from('comment').where('noteId', '=', parent.id)

Newbie question
How we know that `parent.id` is the way to filter comments. 
How this will work with a relationship table - it might be tricky.

**Comment by wtrocki** [comment]
Need some time to play with this to check how it works with all the cases. 
Generally, the code looks sane. I will suggest to not block you and go 

We might badly need integration tests as things can break quite fast with the amount of features we will add so verification could take very long time. 
WDYT?

**Comment by ankitjena** [comment]
> > return context.db.select().from('comment').where('noteId', '=', parent.id)
> 
> Newbie question
> How we know that `parent.id` is the way to filter comments.
> How this will work with a relationship table - it might be tricky.

It's the way we deal with the database. For 1:1 and 1:m we create column. And for how we create these resolvers the approach is known

**Comment by ankitjena** [comment]
> Need some time to play with this to check how it works with all the cases.
> Generally, the code looks sane. I will suggest to not block you and go
> 
> We might badly need integration tests as things can break quite fast with the amount of features we will add so verification could take very long time.
> WDYT?

Yes we should add integration tests to cli

**Comment by wtrocki** [comment]
> How we know that parent.id is the way to filter comments.

I mean here that parent type very often not have `id` as field at all.
We need to get it from relationship field (in fact we have this in directive already)

**Comment by ankitjena** [comment]
> > How we know that parent.id is the way to filter comments.
> 
> I mean here that parent type very often not have `id` as field at all.
> We need to get it from relationship field (in fact we have this in directive already)

Well `id` will be present in database as a column and for the cases we are fetching all the columns from database now. But I see what you mean - that when we use selective querying. 

Didn't quite get the directive reference.

**Comment by wtrocki** [comment]
type may not have id. It is just that our sample model have it :)

**Comment by wtrocki** [comment]
Let's update docs as well

**Comment by ankitjena** [comment]
> type may not have id. It is just that our sample model have it :)

But the postgres table would have

**Comment by wtrocki** [comment]
ah.. ok so this is a bug on postgress. If we specify ID on field it should be added. Lets go with this for now

**Comment by ankitjena** [comment]
> ah.. ok so this is a bug on postgress. If we specify ID on field it should be added. Lets go with this for now

Not a bug exactly. We are creating that in `table.primary`

**Pull request merged**